apps/frontend/src/app/Solver/GOSolver/linearUB.ts

```diff
@@ -0,0 +1,80 @@
+import { OptNode } from "../../Formula/optimization";
+import { cartesian } from "../../Util/Util";
+import { ArtifactsBySlot, MinMax, computeFullArtRange } from "../common";
+import { Linear } from "./BNBSplitWorker";
+import { polyUB } from "./polyUB";
+import { solveLP } from "./solveLP_simplex";
+
+function weightedSum(...entries: readonly (readonly [number, Linear])[]): Linear {
+  const result = { $c: 0 }
+  for (const [weight, entry] of entries)
+    for (const [k, v] of Object.entries(entry))
+      result[k] = (result[k] ?? 0) + weight * v
+  return result
+}
+
+export function linearUB(nodes: OptNode[], arts: ArtifactsBySlot) {
+  const polys = polyUB(nodes, arts)
+  const minMax = computeFullArtRange(arts)
+
+  return polys.map(poly =>
+    weightedSum(...poly.map(mon => {
+      const bounds = mon.terms.map(key => minMax[key])
+      const { w, $c } = lub(bounds)
+      const linboi: Linear = { $c }
+      mon.terms.forEach((key, i) => linboi[key] = w[i])
+      return [mon.$k, linboi] as readonly [number, Linear]
+    }))
+  )
+}
+
+/**
+ * Constructs a linear upper bound for a monomial on a bounded domain using an LP.
+ *
+ * Monomial is assumed to be
+ *    m(x) = x1 * x2 * ... * xn
+ * on bounded domain
+ *    min_1 <= x1 <= max_1
+ *    min_2 <= x2 <= max_2
+ *    ...
+ *    min_n <= xn <= max_n
+ *
+ * @param bounds List of min/max bounds for each xi
+ * @returns A linear function L(x) = w•x + $c
```

Comment:
Let's stick to ASCII (`*`) even if it is slightly less accurate.

Reply:
do you accept `@` as inner product symbol? If not imma do `w . x` because I typically write `w * x` to mean element-wise product

Reply:
Either should be fine, or we can just use `dot(w, x)` if we don't have a commonly used notation.